Source/SwiftLintFramework/Rules/Lint/UnusedDeclarationRule.swift

```diff
@@ -0,0 +1,277 @@
+    func allCursorInfo(compilerArguments: [String]) -> [[String: SourceKitRepresentable]] {
+        guard let path = path, let editorOpen = try? Request.editorOpen(file: self).sendIfNotDisabled() else {
+            return []
+        }
+
+        return syntaxMap.tokens.compactMap { token in
+            guard let kind = SyntaxKind(rawValue: token.type), !syntaxKindsToSkip.contains(kind) else {
+                return nil
+            }
+
+            let offset = Int64(token.offset)
+            let request = Request.cursorInfo(file: path, offset: offset, arguments: compilerArguments)
+            guard var cursorInfo = try? request.sendIfNotDisabled() else {
+                return nil
+            }
+
+            if let acl = File.aclAtOffset(offset, substructureElement: editorOpen) {
+                cursorInfo["key.accessibility"] = acl
+            }
+            cursorInfo["swiftlint.offset"] = offset
+            return cursorInfo
+        }
+    }
+
+    static func declaredUSRs(allCursorInfo: [[String: SourceKitRepresentable]], includePublicAndOpen: Bool)
+        -> [(usr: String, nameOffset: Int)] {
+        return allCursorInfo.compactMap { cursorInfo in
+            return declaredUSRAndOffset(cursorInfo: cursorInfo, includePublicAndOpen: includePublicAndOpen)
+        }
+    }
+
+    static func referencedUSRs(allCursorInfo: [[String: SourceKitRepresentable]]) -> [String] {
+        return allCursorInfo.compactMap(referencedUSR)
+    }
+
+    static func testCaseUSRs(allCursorInfo: [[String: SourceKitRepresentable]]) -> Set<String> {
+        return Set(allCursorInfo.compactMap(testCaseUSR))
+    }
+
+    private static func declaredUSRAndOffset(cursorInfo: [String: SourceKitRepresentable], includePublicAndOpen: Bool)
+        -> (usr: String, nameOffset: Int)? {
+        if let offset = cursorInfo["swiftlint.offset"] as? Int64,
+            let usr = cursorInfo["key.usr"] as? String,
+            let kind = (cursorInfo["key.kind"] as? String).flatMap(SwiftDeclarationKind.init(rawValue:)),
+            !declarationKindsToSkip.contains(kind),
+            let acl = (cursorInfo["key.accessibility"] as? String).flatMap(AccessControlLevel.init(rawValue:)),
+            includePublicAndOpen || [.internal, .private, .fileprivate].contains(acl) {
+            // Skip declarations marked as @IBOutlet, @IBAction or @objc
+            // since those might not be referenced in code, but only dynamically (e.g. Interface Builder)
+            if let annotatedDecl = cursorInfo["key.annotated_decl"] as? String,
+                ["@IBOutlet", "@IBAction", "@objc", "@IBInspectable"].contains(where: annotatedDecl.contains) {
+                return nil
+            }
+
+            // Classes marked as @UIApplicationMain are used by the operating system as the entry point into the app.
+            if let annotatedDecl = cursorInfo["key.annotated_decl"] as? String,
+                annotatedDecl.contains("@UIApplicationMain") {
```

Comment:
should we add `@objcMembers` in this same spot?

Reply:
Good question. You could still have unused `@objcMembers` classes though.

I'm inclined to still treat those as violations until someone requests otherwise.